pkg/storegateway/indexheader/index/symbols_test.go

```diff
@@ -25,7 +27,91 @@ func TestMain(m *testing.M) {
 	test.VerifyNoLeakTestMain(m)
 }
 
-func TestSymbols(t *testing.T) {
+func TestSymbolsV1(t *testing.T) {
+	const (
+		numSymbols       = 100
+		symbolsTableSize = 294         // excludes leading length field and the trailing crc32
+		tableLengthSize  = 4           // number of bytes which store the length of the symbols table
+		filePrefix       = "something" // simulate the file containing something other than the symbols table
+	)
+	var (
+		sortedSymbols    [numSymbols]string
+		sortedSymbolRefs [numSymbols]uint32
+		refToSymbol      = make(map[uint32]string, numSymbols)
+	)
+
+	for i := 0; i < numSymbols; i++ {
+		sortedSymbols[i] = strconv.Itoa(i)
+	}
+	// We must write the symbols sorted in the table, the TSDB index does the same.
+	slices.Sort(sortedSymbols[:])
+
+	buf := encoding.Encbuf{}
+	buf.PutUvarintStr(filePrefix)
+	symbolsStart := buf.Len()
+	buf.PutBE32int(symbolsTableSize)
+	buf.PutBE32int(numSymbols)
+
+	for i, symbol := range sortedSymbols {
+		// The symbol reference in v1 is the offset of the string (and its varint length) in the index header file,
+		// so we use the length of the buffer as the reference (i.e. length of the file).
+		symbolRef := uint32(buf.Len())
+		refToSymbol[symbolRef] = symbol
+		sortedSymbolRefs[i] = symbolRef
+		buf.PutUvarintStr(symbol)
+	}
+	checksum := crc32.Checksum(buf.Get()[symbolsStart+tableLengthSize:], castagnoliTable)
+	buf.PutBE32(checksum)
+
+	dir := t.TempDir()
+	filePath := path.Join(dir, "index")
+	require.NoError(t, os.WriteFile(filePath, buf.Get(), 0700))
+
+	reg := prometheus.NewPedanticRegistry()
+	df := streamencoding.NewDecbufFactory(filePath, 0, log.NewNopLogger(), streamencoding.NewDecbufFactoryMetrics(reg))
+	s, err := NewSymbols(df, index.FormatV1, symbolsStart)
+	require.NoError(t, err)
+
+	// We store only 4 offsets to symbols.
+	require.Len(t, s.offsets, 4)
+
+	t.Run("Lookup", func(t *testing.T) {
+		for ref, symbol := range refToSymbol {
+			s, err := s.Lookup(ref)
+			require.NoError(t, err)
+			require.Equal(t, symbol, s)
+		}
+
+		// Look beyond the last symbol
+		_, err = s.Lookup(uint32(symbolsTableSize + len(filePrefix) + 1))
+		require.Error(t, err)
```

Comment:
Can we be more specific about which error we expect here? (same comment applies to other `require.Error` calls)

Reply:
it required more massaging than I expected, but take a look at https://github.com/grafana/mimir/pull/5103/commits/508efb9ba36d333f300821cf66059cc69b83b397